coderd/wsbuilder/wsbuilder.go

```diff
@@ -611,19 +557,33 @@ func (b *Builder) getTemplateVersionParameters() ([]database.TemplateVersionPara
 	return tvp, nil
 }
 
-func (b *Builder) getLastParameterValues() ([]database.ParameterValue, error) {
-	if b.lastParameterValues != nil {
-		return *b.lastParameterValues, nil
+// verifyNoLegacyParameters verifies that initiator can't start the workspace build
+// if it uses legacy parameters (database.ParameterValues).
+func (b *Builder) verifyNoLegacyParameters() error {
+	if b.verifyNoLegacyParametersOnce {
+		return nil
+	}
+	b.verifyNoLegacyParametersOnce = true
+
+	// Block starting the workspace with legacy parameters.
+	if b.trans != database.WorkspaceTransitionStart {
+		return nil
 	}
+
 	pv, err := b.store.ParameterValues(b.ctx, database.ParameterValuesParams{
 		Scopes:   []database.ParameterScope{database.ParameterScopeWorkspace},
 		ScopeIds: []uuid.UUID{b.workspace.ID},
 	})
-	if err != nil && !xerrors.Is(err, sql.ErrNoRows) {
-		return nil, xerrors.Errorf("get workspace %w parameter values: %w", b.workspace.ID, err)
+	if xerrors.Is(err, sql.ErrNoRows) {
+		return nil
 	}
-	b.lastParameterValues = &pv
-	return pv, nil
+	if err != nil {
+		return xerrors.Errorf("get workspace %w parameter values: %w", b.workspace.ID, err)
+	}
+	if len(pv) > 0 {
+		return xerrors.Errorf("Legacy parameters are not supported anymore, delete the workspace and the related template.")
```

Comment:
Including this check will make it impossible to start workspaces that were migrated.

Prior to this change, we would have deleted ParameterValues only if we were overwriting them with new ones.  We have never had code that cleans out ParameterValues that are no longer used in the template version.

Reply:
> Including this check will make it impossible to start workspaces that were migrated.

Sorry, did you mean "were" migrated or "were not" migrated?

"were not" - yes, workspaces will be bricked, but owners have been warned in the previous release about the deprecation. It was acknowledged with @bpmct.

"were" - I assume that template revisions not using parameter values don't keep records for old parameter values in the database. Has this behavior changed with `wsbuilder`?

Reply:
I mean "were"

ParameterValues scoped to the workspace don't get cleaned up, even if the template changes such that the template no longer consumes those parameters.  We only delete if we're about to write a new one with the same name. This behavior has not changed with `wsbuilder`: https://github.com/coder/coder/blame/622456faf8528d2000e28a683e765cf5f04efda5/coderd/workspacebuilds.go#L579

Reply:
We had a 1-1 with Spike around this issue.

@bpmct Is it a hard requirement to smoothly delete the workspace which uses legacy parameters or could users depend on the "remove orphan" mode? It looks like for the "smooth delete", we need to pull legacy parameters from the database, and this will force us to keep more legacy code.

Actually we're talking about removing all code related to legacy parameters, including the gentle warning blocking users from starting the "legacy" workspace.